### PR TITLE
Evaluate string expression BODMAS (stack) code added

### PR DESCRIPTION
**Pull request by AayushBangroo**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)
Yes
### Description
Corresponding Issue no. : #1719 
(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)

### CODE REVIEW

**Comment by AayushBangroo** [comment]
review @HarshCasper

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+Time Complexity: O(n)
+Space Complexity: O(n)
+*/
+#include<iostream>
+#include<string>
+#include<stack>
+using namespace std;
+
+//The user can provide the same input (2+3)*5 as (2 +3)*5, (  2+3) *5, (2 + 3) * 5 and so on.
+//Therefore to make the algorithm function properly trim the unnecessary spaces from the string
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+			if(operators.empty()){
+				operators.push(token[i]);
+			}
+			else{
+				//check precedence
+				//During expression evaluation NOTE that a strictly higher precedence operator can 
+				//sit on top of the current operator in 'operators' stack
+				//otherwise we need to evaluate the expression until the precedence of current operator in 'operators' stack becomes  
+				//strictly lower or 'operators' stack becomes empty
+				//if precedence of current symbol is more than top of 'operators' stack then simply push it
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Try to make Comments compact and easy to understand.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Try to remove extra lines of code.

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp (hunk 2 above)
Comment (AayushBangroo) [reply C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: updated

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+					//After the above while loop is over either the 'operators' stack is empty or the current operator in  
+					//'operators' stack has lower precedence than operator of given string so we can safely push it
+					operators.push(token[i]);
+				}
+			}
+		}
+	}
+	
+	//After the string is parsed if operators stack is still not empty this means the expression is not fully calculated. 
+	//Note that size of operators stack will always be less than size of 'values' stack.
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+						
+						char symbol = operators.top();
+						operators.pop();
+						
+						int calculated_value = calculate(val1,val2,symbol);
+						
+						values.push(calculated_value);
+					}
+					//After the above while loop is over either the 'operators' stack is empty or the current operator in  
+					//'operators' stack has lower precedence than operator of given string so we can safely push it
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		}
+		
+		else{//operator encountered
+			//if 'operators' stack is empty simply push the operator
+			if(operators.empty()){
+				operators.push(token[i]);
+			}
+			else{
+				//check precedence
+				//if precedence of current symbol is more than symbol at top of 'operators' stack then simply push it
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+				int val1 = values.top();
+				values.pop();
+
+				//get the operator from operator stack
+				char symbol = operators.top();
+				operators.pop();
+
+				//we now have 2 operands and one operator therefore we can calculate the value and store the calculated value back into 
+				//the 'values' stack for futher evaluation
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		else if(token[i]=='('){
+			operators.push('(');
+		}
+
+		//If a closing paranthesis is encountered we have to evaluate the expression until an opening paranthesis is encountered
+		//Assuming the expression is valid we are guaranteed to obtain an opening paranthesis '('
+		else if(token[i]==')'){
+			while(!operators.empty() && operators.top()!='('){
+				//since stack stores values in Last In First Out(LIFO) fashion the top value in the 'values' stack is the second operand 
+				//e.g if expression is (4+5), the top value in stack will be 5 which is the second operand and 4 is the first operand
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+			i--;//decrement i to avoid incrementing it 2 times as outer for loop also increments it
+	
+		}
+		//if opening paranthesis is encontered simply push it in 'operators' stack
+		else if(token[i]=='('){
+			operators.push('(');
+		}
+
+		//If a closing paranthesis is encountered we have to evaluate the expression until an opening paranthesis is encountered
+		//Assuming the expression is valid we are guaranteed to obtain an opening paranthesis '('
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp]: Convert to Multi-Line Comments

File: C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+		return num1+num2;
+	}
+	else if(symbol=='-'){
+		return num1-num2;
+	}
+	return -1;
+}
+
+//Two stacks are used. 'values' stack is used to store the operands
+//'operators' stack is used to hold the operators
```
Comment (HarshCasper) [inline C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp:59]: Convert to Multi-Line Comments

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Please use the comments when apt. The Code should be self explanatory and easy to understand, so don't use comments where unnecessary.

Once the changes are done, please squash the commits and push the changes

**Comment by HarshCasper** [comment]
Are you still working on this @AayushBangroo ?

**Comment by AayushBangroo** [comment]
Yes@HarshCasper

**Pull request merged**